File: 2015/03/part_1.py
#  Usage: From root folder
#  $ python 2015/03/part_1.py 2015/03/input

import sys


class Position:
    def __init__(self):
        self.x, self.y = 0, 0

    def get_key(self):
        return f"{self.x}{self.y}"

    def move(self, direction):
        if direction == "<":
            self.x -= 1
        elif direction == ">":
            self.x += 1
        elif direction == "^":
            self.y += 1
        elif direction == "v":
            self.y -= 1


if __name__ == "__main__":
    with open(sys.argv[1]) as file:
        santa = Position()
        houses = {santa.get_key(): True}
        for directions in file.readlines():
            for direction in directions.strip():
                santa.move(direction)
                houses[santa.get_key()] = True
            print(len(houses))

File: 2015/03/part_1.rb
# Usage: From root folder
# $ ruby 2015/03/part_1.rb 2015/03/input

class Position
  def initialize
    @x, @y = [0, 0]
  end

  def get_key
    "#{@x}#{@y}"
  end

  def move direction
    case direction
    when "<"
      @x-=1
    when ">"
      @x+=1
    when "^"
      @y+=1
    when "v"
      @y-=1
    end
  end
end

File.open(ARGV[0]).readlines.each do |directions|
  santa  = Position.new
  houses = {santa.get_key => true}

  directions.split("").each do |direction|
    santa.move direction
    houses[santa.get_key] = true
  end

  puts(houses.size)
end
